hw/ip/spi_host/doc/_index.md

```diff
@@ -0,0 +1,633 @@
+---
+title: "SPI_HOST HWIP Technical Specification"
+---
+
+<!-- NOTE: Many inline comments in this draft marked "QFR" or Question For Reviewers-->
+<!-- QFR comments to be removed in final squash & merge -->
+
+# Overview
+
+This document specifies SPI_HOST hardware IP (HWIP) functionality.
+This module conforms to the [Comportable guideline for peripheral functionality.]({{< relref "doc/rm/comportability_specification/index.md" >}})
+See that document for integration overview within the broader top-level system.
+
+## Features
+
+- Hardware control for remote devices using the Serial Peripheral Interface (SPI)
+- Primarily designed for serial NOR flash devices such as the [Winbond W25X20CL](https://www.winbond.com/resource-files/w25x20cl_revf%2020150806.pdf)
+- Chip Select support for up to 8 devices
+- Support for Standard SPI, Dual SPI or Quad SPI commands
+- RX and TX data held in separate 512 byte FIFOs
```

Comment:
@martin-lueker do you think it's feasible to.. instead of having 2 512B fifos, 2 256B fifos (or even slightly bigger) that can be configured to be 1 R / 1 W, or both reads and both writes?

Especially if we think we're primarily supporting serial flash ... most of the time, the transaction is not full duplex.  So only one out of the two FIFOs would really be used.  If we can have flexible assignment of FIFOs to function, it also allows for software to ping / pong.

This is I believe a nice to have, but it might be good to think through. I think the original one pager might have also discussed this.

Reply:
Another thing we probably want to check with software is whether there is desire for a "start index".

Reply:
I suppose one could definitely shrink it to 256+n.  So there is some savings to be had there.

I understand your point about the half-duplex, but particularly with the two clock domains, I don't think making them dual purpose sounds straight forward to implement.  Nor was this in Shaun's one-pager.

Reply:
You'll have to explain what you mean by a "start index".  @gkelly, @a-pronin, do you have any special requests along this line?

Reply:
so it turns out the start index is described elsewhere i believe (let me know if i'm wrong)... the main point i was trying to get across is that a lot of the protocol isn't full duplex.. so when you are transmitting, there's actually nothing interesting to be received.  So it may make sense to explicitly indicate "when" you need to actually start receiving. 

This is only useful if the maximum payload size is "exactly" the size of your FIFO.  Since that means a 512B entry for example wastes a few bytes early on during command / address / dummy phases and can't actually receive the whole payload.  The firmware then needs to cut down to some other size etc.  This is also somewhat related to the flexible buffer size, which is originally what Randall described (though to me not a requirement).  It allows them to make skew the buffer in one direction (big RX, tiny TX or vice versa).

This is not super critical, but something to consider.

Reply:
Oh, I see.  Sure, we can set up all commands (Standard included) to only catch the interesting bytes.   That is super easy.

Reply:
Bidirectional async fifos on the other hand.  I'd really like to avoid that if I can get away with it.   I know that SPI_DEVICE is pretty flexible in that regard, and offloads a lot of complexity to firmware, but my first guess is that it would take a pretty big bump in effort to adapt that to SPI_HOST I think.

Reply:
ACTIONS TAKEN:
- FIFO sizes reduced to 264 bytes (256 + 8). Since async fifos are limited to powers of two this will need to be an async FIFO with a tiny sync FIFO add-on.   Not a problem though.
- All commands are now half-duplex by default.  If I understand correctly I think this settles the "start-index" concern.

Reply:
@tjaychen, is there anything else that we can do for this point?

Reply:
sounds good. Just wanted to make sure, just because most commands are half duplex, doesn't mean ALL of them are :)
So as long it's possible to configure which, i think that sounds good.

Reply:
Great.